File: sdk/python/omoios/resources/sessions.py
"""Sessions resource — spec §03 primary SDK surface.

Methods mirror the backend's `/api/v1/sessions/*` routes one-to-one:

    client.sessions.create(...)            POST   /api/v1/sessions
    client.sessions.get(id)                GET    /api/v1/sessions/{id}
    client.sessions.list(...)              GET    /api/v1/sessions   (auto-paginate)
    client.sessions.cancel(id)             DELETE /api/v1/sessions/{id}
    client.sessions.reply(id, text)        POST   /api/v1/sessions/{id}/messages
    client.sessions.fork(id, seq, prompt)  POST   /api/v1/sessions/{id}/fork
    client.sessions.share(id, grants)      POST   /api/v1/sessions/{id}/share
    client.sessions.artifacts(id)          GET    /api/v1/sessions/{id}/artifacts
    client.sessions.events(id, ...)        GET    /api/v1/sessions/{id}/events   (SSE)
    client.sessions.connect(id, token)     WS     /api/v1/sessions/{id}/ws

Four primitive interaction patterns from spec §18 are all expressible here:

    A. Fire-and-forget:  s = await client.sessions.create(...)
    B. Sync wait:        async for e in client.sessions.events(s.id): ...
    C. Live stream:      same as B, render each event as it arrives
    D. Multiplayer:      ch = client.sessions.connect(s.id, jwt); ch.on(...)
"""

from __future__ import annotations

import json
import uuid
from typing import Any, AsyncIterator, Awaitable, Callable, Dict, List, Optional
from urllib.parse import urljoin

from omoios.resources.base import BaseResource
from omoios.types import (
    Artifact,
    Event,
    Grant,
    Session,
)


class SessionsResource(BaseResource):
    """Session lifecycle + streaming (spec §03, §07, §18)."""

    # ─── core lifecycle ─────────────────────────────────────────────────────

    async def create(
        self,
        *,
        prompt: str,
        workspace_id: Optional[str] = None,
        environment_id: Optional[str] = None,
        github_repo: Optional[str] = None,
        share_with: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        idempotency_key: Optional[str] = None,
        cancel_scope: Optional[Any] = None,
    ) -> Session:
        """Create a new session (spec §03).

        Either `workspace_id` or `github_repo` must be supplied. When only
        `github_repo="owner/repo"` is given, the backend auto-binds a
        workspace to that repo within the caller's org (mirrors the ticket
        auto-project behaviour on the dashboard side).

        `idempotency_key` is auto-generated if omitted; pass your own to dedup
        retries across network failures. Same key + same body = same session.
        """
        if not workspace_id and not github_repo:
            raise ValueError("Provide either `workspace_id` or `github_repo`")

        body: Dict[str, Any] = {"prompt": prompt}
        if workspace_id is not None:
            body["workspace_id"] = workspace_id
        if environment_id is not None:
            body["environment_id"] = environment_id
        if github_repo is not None:
            body["github_repo"] = github_repo
        if share_with:
            body["share_with"] = list(share_with)
        if metadata is not None:
            body["metadata"] = metadata

        headers = {
            "Idempotency-Key": idempotency_key or str(uuid.uuid4()),
        }
        response = await self._client._request(
            "POST",
            "/api/v1/sessions",
            json=body,
            headers=headers,
            cancel_scope=cancel_scope,
        )
        return Session(**response.json())

    async def get(
        self,
        session_id: str,
        *,
        cancel_scope: Optional[Any] = None,
    ) -> Session:
        """Fetch a session by id."""
        response = await self._client._request(
            "GET",
            f"/api/v1/sessions/{session_id}",
            cancel_scope=cancel_scope,
        )
        return Session(**response.json())

    async def list(
        self,
        *,
        status: Optional[str] = None,
        phase_id: Optional[str] = None,
        ticket_id: Optional[str] = None,
        page_size: int = 100,
        cancel_scope: Optional[Any] = None,
    ) -> AsyncIterator[Session]:
        """List sessions with auto-pagination (spec §09).

        Yields sessions one at a time. Under the hood this pages using the
        backend's `limit`/`offset` params. For the eager `list[Session]`
        shape, use `list_all()` instead.
        """
        offset = 0
        while True:
            params: Dict[str, Any] = {"limit": page_size, "offset": offset}
            if status is not None:
                params["status"] = status
            if phase_id is not None:
                params["phase_id"] = phase_id
            if ticket_id is not None:
                params["ticket_id"] = ticket_id

            response = await self._client._request(
                "GET",
                "/api/v1/sessions",
                params=params,
                cancel_scope=cancel_scope,
            )
            items = response.json()
            if not isinstance(items, list):
                items = items.get("items", [])

            for item in items:
                yield Session(**item)

            if len(items) < page_size:
                return
            offset += page_size

    async def list_all(self, **kwargs: Any) -> List[Session]:
        """Eager variant of `list()` — collects all pages into a list."""
        return [s async for s in self.list(**kwargs)]

    async def cancel(
        self,
        session_id: str,
        *,
        cancel_scope: Optional[Any] = None,
    ) -> Dict[str, Any]:
        """Cancel a running session. Idempotent."""
        response = await self._client._request(
            "DELETE",
            f"/api/v1/sessions/{session_id}",
            cancel_scope=cancel_scope,
        )
        return response.json()

    # ─── spec §03 lifecycle actions ─────────────────────────────────────────

    async def reply(
        self,
        session_id: str,
        text: str,
        *,
        cancel_scope: Optional[Any] = None,
    ) -> None:
        """Send a follow-up prompt mid-session. Non-blocking."""
        await self._client._request(
            "POST",
            f"/api/v1/sessions/{session_id}/messages",
            json={"text": text},
            cancel_scope=cancel_scope,
        )

    async def fork(
        self,
        session_id: str,
        from_seq: int,
        prompt: str,
        *,
        cancel_scope: Optional[Any] = None,
    ) -> Session:
        """Branch a session at event `from_seq` with a new prompt."""
        response = await self._client._request(
            "POST",
            f"/api/v1/sessions/{session_id}/fork",
            json={"from_seq": from_seq, "prompt": prompt},
            cancel_scope=cancel_scope,
        )
        return Session(**response.json())

    async def share(
        self,
        session_id: str,
        grants: List[Grant],
        *,
        cancel_scope: Optional[Any] = None,
    ) -> Dict[str, Any]:
        """Grant ACL roles on a session (spec §07)."""
        body = {
            "grants": [
                {"user_id": str(g.user_id), "role": g.role} for g in grants
            ]
        }
        response = await self._client._request(
            "POST",
            f"/api/v1/sessions/{session_id}/share",
            json=body,
            cancel_scope=cancel_scope,
        )
        return response.json()

    async def artifacts(
        self,
        session_id: str,
        *,
        cancel_scope: Optional[Any] = None,
    ) -> List[Artifact]:
        """List artifacts produced by this session."""
        response = await self._client._request(
            "GET",
            f"/api/v1/sessions/{session_id}/artifacts",
            cancel_scope=cancel_scope,
        )
        rows = response.json()
        return [Artifact(**row) for row in rows]

    # ─── spec §03 streaming ─────────────────────────────────────────────────

    async def events(
        self,
        session_id: str,
        *,
        last_event_id: Optional[str] = None,
    ) -> AsyncIterator[Event]:
        """Stream session events as a native async iterator.

        Uses `httpx-sse` under the hood — robust SSE framing, `Last-Event-ID`
        header for resume. The iterator closes cleanly when the server
        finishes (replay exhausted + no Redis) or when the caller breaks
        out of the loop.

        Example (spec §18 Pattern B — sync wait):

            async for evt in client.sessions.events(session_id):
                if evt.type == "session.succeeded":
                    break
        """
        # Import here so the SDK imports cleanly even when httpx-sse is
        # absent — callers only pay the dependency when they stream events.
        try:
            from httpx_sse import aconnect_sse
        except ImportError as exc:  # pragma: no cover
            raise RuntimeError(
                "Streaming events requires `httpx-sse`. "
                "Install with `pip install omoios[sse]` or "
                "`pip install httpx-sse`."
            ) from exc

        headers = {
            **self._client._headers(),
            "Accept": "text/event-stream",
        }
        if last_event_id:
            headers["Last-Event-ID"] = last_event_id

        url = urljoin(
            self._client.base_url + "/",
            f"api/v1/sessions/{session_id}/events",
        )

        import time

        stream_path = f"/api/v1/sessions/{session_id}/events"
        self._client._emit_telemetry(
            {"kind": "stream_open", "path": stream_path}
        )
        started = time.perf_counter()
        frames = 0
        try:
            async with aconnect_sse(
                self._client._http, "GET", url, headers=headers
            ) as es:
                async for sse in es.aiter_sse():
                    if not sse.data:
                        continue
                    try:
                        payload = json.loads(sse.data)
                    except (ValueError, TypeError):
                        continue
                    frames += 1
                    yield Event(**payload)
        finally:
            self._client._emit_telemetry(
                {
                    "kind": "stream_close",
                    "path": stream_path,
                    "frames_received": frames,
                    "duration_ms": (time.perf_counter() - started) * 1000,
                }
            )

    def connect(
        self, session_id: str, user_token: Optional[str] = None
    ) -> "SessionChannel":
        """Open a multiplayer WebSocket channel (spec §07).

        `user_token` defaults to the client's active JWT; pass a different
        one for a delegated session.
        """
        token = user_token or self._client.jwt_token or self._client.api_key
        return SessionChannel(
            client=self._client, session_id=session_id, token=token or ""
        )


# ─── multiplayer channel ────────────────────────────────────────────────────


class SessionChannel:
    """Thin wrapper around a per-session WebSocket (spec §07).

    Usage:

        ch = client.sessions.connect(session_id, user_jwt)
        ch.on("participant.joined", on_join)
        ch.on("session.message", on_msg)
        await ch.open()
        await ch.send({"type": "message.send", "data": {"text": "hi"}})
        await ch.close()

    Or as an async context manager:

        async with client.sessions.connect(session_id, jwt) as ch:
            ch.on("*", print)
            await asyncio.sleep(30)
    """

    def __init__(self, *, client, session_id: str, token: str):
        self._client = client
        self._session_id = session_id
        self._token = token
        self._ws = None
        self._ws_ctx = None  # httpx-ws async-context-manager handle
        self._reader_task = None
        self._handlers: Dict[str, List[Callable[[Dict[str, Any]], Any]]] = {}
        self._star_handlers: List[Callable[[Dict[str, Any]], Any]] = []

    # ── event subscriptions ────────────────────────────────────────────────

    def on(
        self,
        event_type: str,
        fn: Callable[[Dict[str, Any]], Any],
    ) -> None:
        """Register a handler for a message type. Use `"*"` to catch all."""
        if event_type == "*":
            self._star_handlers.append(fn)
        else:
            self._handlers.setdefault(event_type, []).append(fn)

    # ── lifecycle ─────────────────────────────────────────────────────────

    async def open(self) -> "SessionChannel":
        """Connect the WebSocket and start the read loop.

        Uses `httpx-ws` so the channel rides on the same httpx transport as
        the rest of the SDK — shared connection pool, shared timeouts, and
        the same auth-header plumbing. No separate `websockets` dependency.
        """
        import asyncio

        try:
            from httpx_ws import aconnect_ws
        except ImportError as exc:  # pragma: no cover
            raise RuntimeError(
                "WebSocket multiplayer requires `httpx-ws`. "
                "Install with `pip install omoios[ws]` or "
                "`pip install httpx-ws`."
            ) from exc

        # Build ws:// or wss:// URL from the client's base URL.
        base = self._client.base_url
        ws_scheme = "wss" if base.startswith("https") else "ws"
        url = (
            f"{ws_scheme}://{base.split('://', 1)[1]}"
            f"/api/v1/sessions/{self._session_id}/ws"
            f"?token={self._token}"
        )

        # `aconnect_ws` returns an async context manager. Enter it manually
        # so the socket stays open for the lifetime of this channel; we
        # release in `close()`.
        self._ws_ctx = aconnect_ws(url, self._client._http)
        self._ws = await self._ws_ctx.__aenter__()
        self._reader_task = asyncio.create_task(self._read_loop())
        self._client._emit_telemetry(
            {
                "kind": "stream_open",
                "path": f"/api/v1/sessions/{self._session_id}/ws",
            }
        )
        import time as _time
        self._opened_at = _time.perf_counter()
        self._frames_received = 0
        return self

    async def send(self, message: Dict[str, Any]) -> None:
        """Send one message frame (spec §07 shapes: message.send, cursor.moved)."""
        if self._ws is None:
            raise RuntimeError("Channel is not open; call .open() first")
        await self._ws.send_text(json.dumps(message))

    async def close(self) -> None:
        """Close the WebSocket and tear down the reader task."""
        import asyncio as _asyncio

        # Cancel the reader first so it doesn't try to read from a closed
        # socket during teardown. CancelledError is a BaseException in
        # Python 3.8+, not Exception — catch both so the cancel doesn't
        # bubble out of close().
        if self._reader_task is not None:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except (_asyncio.CancelledError, Exception):  # noqa: BLE001
                pass
            self._reader_task = None

        if self._ws is not None and self._ws_ctx is not None:
            try:
                await self._ws_ctx.__aexit__(None, None, None)
            except (_asyncio.CancelledError, Exception):  # noqa: BLE001
                pass
            self._ws = None
            self._ws_ctx = None

        # Telemetry close — fires once per channel regardless of how we
        # got here (explicit close or context-manager exit).
        if getattr(self, "_opened_at", None) is not None:
            import time as _time
            self._client._emit_telemetry(
                {
                    "kind": "stream_close",
                    "path": f"/api/v1/sessions/{self._session_id}/ws",
                    "frames_received": getattr(self, "_frames_received", 0),
                    "duration_ms": (_time.perf_counter() - self._opened_at) * 1000,
                }
            )
            self._opened_at = None

    async def __aenter__(self) -> "SessionChannel":
        return await self.open()

    async def __aexit__(self, *exc: Any) -> None:
        await self.close()

    # ── internals ──────────────────────────────────────────────────────────

    async def _read_loop(self) -> None:
        """Dispatch inbound frames to registered handlers."""
        assert self._ws is not None
        try:
            while True:
                raw = await self._ws.receive_text()
                try:
                    frame = json.loads(raw)
                except (ValueError, TypeError):
                    continue
                self._frames_received = getattr(self, "_frames_received", 0) + 1
                # Fire star handlers regardless of shape.
                for handler in self._star_handlers:
                    _maybe_await(handler, frame)
                # Fire type-specific handlers if the frame has a type field.
                msg_type = frame.get("type")
                if msg_type:
                    for handler in self._handlers.get(msg_type, []):
                        _maybe_await(handler, frame)
        except Exception:  # noqa: BLE001 — socket closed or transport error
            return


def _maybe_await(fn: Callable[[Any], Any], arg: Any) -> None:
    """Call `fn(arg)`. If it returns a coroutine, schedule it on the loop."""
    import asyncio

    result = fn(arg)
    if hasattr(result, "__await__"):
        asyncio.create_task(result)  # type: ignore[arg-type]

File: sdk/typescript/src/resources/sessions.ts
/**
 * Sessions resource — spec §03 primary SDK surface.
 *
 * Mirrors the Python SDK method-for-method. All four spec §18 interaction
 * patterns (fire-and-forget / sync wait / live stream / multiplayer) are
 * expressible via this one resource.
 */

import type { OmoiOSClient } from '../client.js';
import type {
  ChannelMessage,
  CreateSessionRequest,
  Event as SessionEvent,
  ForkRequest,
  Grant,
  Session,
  ShareRequest,
} from '../types.js';

// Node's UUID isn't in the global scope; use a tiny inline generator so we
// don't drag `uuid` into the SDK's production dependencies.
function randomIdempotencyKey(): string {
  // RFC4122-shaped random string; good enough for dedup keys.
  const bytes = new Uint8Array(16);
  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
  }
  bytes[6] = (bytes[6]! & 0x0f) | 0x40;
  bytes[8] = (bytes[8]! & 0x3f) | 0x80;
  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

export class SessionsResource {
  constructor(private readonly client: OmoiOSClient) {}

  // ─── core lifecycle ──────────────────────────────────────────────────────

  /**
   * Create a new session (spec §03). Returns in <200ms.
   *
   * Either `workspaceId` or `githubRepo` must be supplied along with
   * `prompt`. When only `githubRepo="owner/repo"` is given the backend
   * auto-binds a workspace in the caller's org (mirrors the ticket
   * auto-project pattern).
   *
   * Pass `idempotencyKey` to dedup retries across network failures; omit
   * for an auto-generated key.
   */
  async create(
    params: {
      prompt: string;
      workspaceId?: string;
      environmentId?: string;
      githubRepo?: string;
      shareWith?: string[];
      metadata?: Record<string, unknown>;
      idempotencyKey?: string;
      signal?: AbortSignal;
    },
  ): Promise<Session> {
    if (!params.workspaceId && !params.githubRepo) {
      throw new Error(
        'sessions.create: either `workspaceId` or `githubRepo` is required',
      );
    }

    const body: CreateSessionRequest = { prompt: params.prompt };
    if (params.workspaceId !== undefined) body.workspace_id = params.workspaceId;
    if (params.environmentId !== undefined)
      body.environment_id = params.environmentId;
    if (params.githubRepo !== undefined) body.github_repo = params.githubRepo;
    if (params.shareWith !== undefined) body.share_with = params.shareWith;
    if (params.metadata !== undefined) body.metadata = params.metadata;

    const response = await this.client._request('POST', '/api/v1/sessions', {
      body: JSON.stringify(body),
      headers: {
        'Idempotency-Key': params.idempotencyKey ?? randomIdempotencyKey(),
      },
      signal: params.signal,
    });
    return (await response.json()) as Session;
  }

  /** Fetch a session by id. */
  async get(
    sessionId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<Session> {
    const response = await this.client._request(
      'GET',
      `/api/v1/sessions/${sessionId}`,
      { signal: options.signal },
    );
    return (await response.json()) as Session;
  }

  /**
   * Auto-paginating list iterator (spec §09).
   *
   * Example:
   *
   *     for await (const s of client.sessions.list({ status: 'running' })) {
   *       console.log(s.id);
   *     }
   */
  async *list(
    params: {
      status?: string;
      phase_id?: string;
      ticket_id?: string;
      pageSize?: number;
      signal?: AbortSignal;
    } = {},
  ): AsyncIterable<Session> {
    const pageSize = params.pageSize ?? 100;
    let offset = 0;
    while (true) {
      const searchParams: Record<string, string> = {
        limit: String(pageSize),
        offset: String(offset),
      };
      if (params.status) searchParams.status = params.status;
      if (params.phase_id) searchParams.phase_id = params.phase_id;
      if (params.ticket_id) searchParams.ticket_id = params.ticket_id;

      const response = await this.client._request(
        'GET',
        '/api/v1/sessions',
        { searchParams, signal: params.signal },
      );
      const parsed = (await response.json()) as Session[] | { items: Session[] };
      const items = Array.isArray(parsed) ? parsed : parsed.items ?? [];

      for (const item of items) yield item;

      if (items.length < pageSize) return;
      offset += pageSize;
    }
  }

  /** Eager variant of `list()` — collects every page into an array. */
  async listAll(params: Parameters<SessionsResource['list']>[0] = {}): Promise<Session[]> {
    const out: Session[] = [];
    for await (const s of this.list(params)) out.push(s);
    return out;
  }

  /** Cancel a running session. Idempotent. */
  async cancel(
    sessionId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<Record<string, unknown>> {
    const response = await this.client._request(
      'DELETE',
      `/api/v1/sessions/${sessionId}`,
      { signal: options.signal },
    );
    return (await response.json()) as Record<string, unknown>;
  }

  // ─── spec §03 lifecycle actions ──────────────────────────────────────────

  /** Send a follow-up prompt mid-session. Non-blocking. */
  async reply(
    sessionId: string,
    text: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<void> {
    await this.client._request(
      'POST',
      `/api/v1/sessions/${sessionId}/messages`,
      { body: JSON.stringify({ text }), signal: options.signal },
    );
  }

  /** Branch a session at event `fromSeq` with a new prompt. */
  async fork(
    sessionId: string,
    fromSeq: number,
    prompt: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<Session> {
    const body: ForkRequest = { from_seq: fromSeq, prompt };
    const response = await this.client._request(
      'POST',
      `/api/v1/sessions/${sessionId}/fork`,
      { body: JSON.stringify(body), signal: options.signal },
    );
    return (await response.json()) as Session;
  }

  /** Grant ACL roles on a session (spec §07). */
  async share(
    sessionId: string,
    grants: Grant[],
    options: { signal?: AbortSignal } = {},
  ): Promise<Record<string, unknown>> {
    const body: ShareRequest = { grants };
    const response = await this.client._request(
      'POST',
      `/api/v1/sessions/${sessionId}/share`,
      { body: JSON.stringify(body), signal: options.signal },
    );
    return (await response.json()) as Record<string, unknown>;
  }

  /** List artifacts produced by this session. */
  async artifacts(
    sessionId: string,
    options: { signal?: AbortSignal } = {},
  ): Promise<Array<Record<string, unknown>>> {
    const response = await this.client._request(
      'GET',
      `/api/v1/sessions/${sessionId}/artifacts`,
      { signal: options.signal },
    );
    return (await response.json()) as Array<Record<string, unknown>>;
  }

  // ─── spec §03 streaming ──────────────────────────────────────────────────

  /**
   * SSE event iterator with `Last-Event-ID` resume (spec §09).
   *
   * Uses the platform `fetch` + `ReadableStream` — no extra deps.
   *
   * Example (Pattern B — sync wait):
   *
   *     for await (const e of client.sessions.events(id)) {
   *       if (e.type === 'session.succeeded') break;
   *     }
   */
  async *events(
    sessionId: string,
    options: { lastEventId?: string; signal?: AbortSignal } = {},
  ): AsyncIterable<SessionEvent> {
    // @ts-expect-error — `_headers` is private on the runtime class but we
    // need it to add auth headers for streaming. Same trick the spec §09
    // example uses.
    const authHeaders: Record<string, string> = this.client._headers?.() ?? {};
    const headers: Record<string, string> = {
      ...authHeaders,
      Accept: 'text/event-stream',
    };
    if (options.lastEventId) headers['Last-Event-ID'] = options.lastEventId;

    const url = new URL(
      `/api/v1/sessions/${sessionId}/events`,
      this.client.baseUrl,
    );

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers,
      signal: options.signal,
    });

    if (!response.ok || !response.body) {
      throw new Error(
        `Event stream failed: ${response.status} ${response.statusText}`,
      );
    }

    const reader = response.body
      .pipeThrough(new TextDecoderStream())
      .getReader();

    let buf = '';
    while (true) {
      const { value, done } = await reader.read();
      if (done) return;
      buf += value;
      // Frames are separated by "\n\n". Split-and-loop handles partial reads.
      let idx: number;
      while ((idx = buf.indexOf('\n\n')) !== -1) {
        const frame = buf.slice(0, idx);
        buf = buf.slice(idx + 2);
        const dataLine = frame
          .split('\n')
          .find((l) => l.startsWith('data: '));
        if (!dataLine) continue;
        const payload = dataLine.slice(6).trim();
        if (!payload) continue;
        try {
          yield JSON.parse(payload) as SessionEvent;
        } catch {
          // Skip malformed frames instead of killing the iterator.
        }
      }
    }
  }

  /** Open a multiplayer WebSocket channel (spec §07). */
  connect(sessionId: string, userToken?: string): SessionChannel {
    const token =
      userToken ?? this.client.jwtToken ?? this.client.apiKey ?? '';
    return new SessionChannel(this.client, sessionId, token);
  }
}

// ─── multiplayer channel ─────────────────────────────────────────────────────

type Handler = (frame: Record<string, unknown>) => void;

export class SessionChannel {
  private ws: WebSocket | null = null;
  private readonly handlers: Map<string, Handler[]> = new Map();
  private readonly starHandlers: Handler[] = [];

  constructor(
    private readonly client: OmoiOSClient,
    private readonly sessionId: string,
    private readonly token: string,
  ) {}

  // ── event subscriptions ────────────────────────────────────────────────

  /** Register a handler for a message type. Use `'*'` to catch all. */
  on(eventType: string, fn: Handler): void {
    if (eventType === '*') {
      this.starHandlers.push(fn);
      return;
    }
    const bucket = this.handlers.get(eventType) ?? [];
    bucket.push(fn);
    this.handlers.set(eventType, bucket);
  }

  // ── lifecycle ──────────────────────────────────────────────────────────

  /** Connect and start listening. Resolves when the socket is open. */
  async open(): Promise<this> {
    const base = this.client.baseUrl;
    const wsScheme = base.startsWith('https') ? 'wss' : 'ws';
    const host = base.split('://', 2)[1];
    const url =
      `${wsScheme}://${host}/api/v1/sessions/${this.sessionId}/ws` +
      `?token=${encodeURIComponent(this.token)}`;

    // Node 18+ exposes global WebSocket via undici; browsers always have it.
    const WSCtor: typeof WebSocket =
      (globalThis as unknown as { WebSocket?: typeof WebSocket }).WebSocket!;
    if (!WSCtor) {
      throw new Error(
        'WebSocket is not available in this runtime. ' +
          'Install `ws` and polyfill globalThis.WebSocket, or use Node 18+.',
      );
    }

    const ws = new WSCtor(url);
    this.ws = ws;

    await new Promise<void>((resolve, reject) => {
      const onOpen = () => {
        ws.removeEventListener('error', onError);
        resolve();
      };
      const onError = (ev: globalThis.Event) => {
        ws.removeEventListener('open', onOpen);
        reject(new Error(`WebSocket connect failed: ${String(ev)}`));
      };
      ws.addEventListener('open', onOpen, { once: true });
      ws.addEventListener('error', onError, { once: true });
    });

    ws.addEventListener('message', (ev) => {
      let frame: Record<string, unknown>;
      try {
        frame =
          typeof ev.data === 'string'
            ? (JSON.parse(ev.data) as Record<string, unknown>)
            : {};
      } catch {
        return;
      }
      for (const h of this.starHandlers) h(frame);
      const msgType = typeof frame.type === 'string' ? frame.type : undefined;
      if (msgType) {
        for (const h of this.handlers.get(msgType) ?? []) h(frame);
      }
    });

    return this;
  }

  /** Send one message frame (spec §07: `message.send`, `cursor.moved`, …). */
  send(message: ChannelMessage): void {
    if (!this.ws) {
      throw new Error('Channel is not open; call .open() first');
    }
    this.ws.send(JSON.stringify(message));
  }

  /** Close the WebSocket. */
  close(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}
